app/lib/backend/storage_rewrite.rb

```diff
@@ -1,106 +1,155 @@
+        service.update_backend_service
+        service.service_tokens.find_each(&ServiceTokenService.method(:update_backend))
       end
+    end
 
-      UsageLimit.find_each do |usage_limit|
-        usage_limit.send(:update_backend_usage_limit)
-        progress.call
-      end
+    class MetricRewriter < Rewriter
+      CLASS = Metric
+      INCLUDE = :parent
+      REWRITER = ->(metric) { metric.sync_backend! }
     end
 
-    def rewrite_provider(id)
-      provider = Account.providers_with_master.find(id)
+    class UsageLimitRewriter < Rewriter
+      CLASS = UsageLimit
+      INCLUDE = { plan: :service }.freeze
+      REWRITER = ->(usage_limit) { usage_limit.update_backend_usage_limit }
+    end
 
-      services = provider.services.includes(:account)
-      cinstances = provider.buyer_applications.includes(:plan, :service)
+    class Processor
+      BATCH_SIZE = 500
 
-      if (cinstance_id = ENV['CINSTANCE_ID'])
-        cinstances = cinstances.where(id: cinstance_id)
-      end
+      attr_reader :include_inactive, :enqueuer, :log_progress, :action
 
-      bought_cinstances = provider.bought_cinstances
+      def initialize(**opts)
+        @include_inactive = opts[:include_inactive] || false
+        @enqueuer = opts[:enqueuer]
+        @action = enqueuer.respond_to?(:call) ? :enqueue : :rewrite
+        @log_progress = opts[:log_progress] || false
+      end
 
-      metrics = Metric.by_provider(provider).includes(:parent)
-      usage_limits = provider.usage_limits.includes(plan: :service)
+      # Execute actual rewriting with Backend, depending on the class.
+      # To be called from the async jobs
+      def rewrite(**kwargs)
+        klass = kwargs[:class_name] || kwargs[:scope]&.klass&.name
+        raise ArgumentError, ':class_name or :scope arguments must be provided' if klass.blank?
 
-      total_count = services.size + cinstances.size + metrics.size + usage_limits.size + bought_cinstances.size
+        rewriter = Backend::StorageRewrite.const_get("#{klass}Rewriter")
+        rewriter.rewrite({ **kwargs, log_progress: log_progress})
+      end
 
-      index = 0
-      progress = -> do
-        percent = ((index + 1) / total_count.to_f) * 100.0
-        yield percent if block_given?
-        index += 1
+      # Schedule all objects for all providers, or execute inline
+      def rewrite_all
+        providers.each do |account|
+          rewrite_provider(account.id)
+        end
+        nil
       end
 
-      services.find_each do |service|
-        service.update_backend_service
-        progress.call
+      # Schedule a single provider or execute inline
+      def rewrite_provider(id)
+        logger.info "#{action} backend storage for provider #{id}..."
 
-        service.service_tokens.find_each(&ServiceTokenService.method(:update_backend))
-      end
+        provider = providers.find_by(id: id)
+        unless provider
+          logger.error("Provider with ID #{id} not found")
```

Comment:
nitpick, if it is normal operation to try to rewrite deleted provider, then maybe use `warn` level. But shouldn't we try to remove data for a provider if it is missing?

Reply:
Hmm, well, if the provider is already deleted from the DB (ID not found), then there is no way we could get a list of its services, applications, limits etc...

We probably do need to have some way to clean up apisonator to remove stuff that has been deleted from porta, but it's out of scope for this PR.

Reply:
so we can use `#warn` in this case?

Reply:
To me `warn` level is when something might be an error but also might be just normal part of operation. Basically we are not sure something is broken or not. Also the use case or deprecations where things are not broken right now but will become in the future.

In this case provider would have been deleted so in such case it would be expected not to find it.

But if I'm wrong and in no occasion we should see this, then it makes more sense to remain `error`.

Reply:
So, yes, there are two main usages:

1. We trigger full rewrite for all providers, in this case indeed not finding a provider is not expected, because we're taking the ID from the loop of all providers. So, "not found" is highly improbable.

2. We trigger full rewrite for a specific provider, by specifying their ID. In this case the user triggering the task can put something invalid, and I think showing an error is correct - like "Hey! Your ID is incorrect, so we can't proceed"

Reply:
I thought sometimes a provider can be gone while rewriting is scheduled. But it is fine with me. Leave it error. Your explanation makes sense.